crates/sargon/src/profile/mfa/security_structures/security_structure_of_factors/security_structure_of_factor_instances.rs

```diff
@@ -0,0 +1,500 @@
+use indexmap::IndexSet;
+
+use crate::prelude::*;
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct SecurityStructureOfFactorInstances {
+    /// The ID of the `SecurityStructureOfFactorSourceIDs` in
+    /// `profile.app_preferences.security.security_structures_of_factor_source_ids`
+    /// which was used to derive the factor instances in this structure. Or rather:
+    /// The id of `SecurityStructureOfFactorSources`.
+    pub security_structure_id: SecurityStructureID,
+
+    /// The structure of factors to use for certain roles, Primary, Recovery
+    /// and Confirmation role.
+    pub matrix_of_factors: MatrixOfFactorInstances,
+
+    /// The authentication signing factor instance which is used to sign
+    /// proof of ownership - aka "True Rola Key". User can select which FactorSource
+    /// to use during Shield Building, but typically most users will user the
+    /// DeviceFactorSource which is default. DerivationPath is in securified
+    /// KeySpace of course.
+    ///
+    /// Non-optional since we can replace it with a new one for entities
+    /// we have recovered during Onboarding Account Recovery Scan for securified
+    /// entities
+    pub authentication_signing_factor_instance:
+        HierarchicalDeterministicFactorInstance,
+}
+
+impl HasFactorInstances for SecurityStructureOfFactorInstances {
+    fn unique_tx_signing_factor_instances(&self) -> IndexSet<FactorInstance> {
+        self.matrix_of_factors.unique_tx_signing_factor_instances()
+    }
+
+    fn unique_all_factor_instances(&self) -> IndexSet<FactorInstance> {
+        let mut instances = self.unique_tx_signing_factor_instances();
+        instances
+            .insert(self.authentication_signing_factor_instance.clone().into());
+        instances
+    }
+}
+
+impl SecurityStructureOfFactorInstances {
+    pub fn new(
+        security_structure_id: SecurityStructureID,
+        matrix_of_factors: MatrixOfFactorInstances,
+        authentication_signing: HierarchicalDeterministicFactorInstance,
+    ) -> Result<Self> {
+        let index_agnostic_path = matrix_of_factors
+            .index_agnostic_path_of_all_tx_signing_factor_instances()?;
+        let entity_kind = index_agnostic_path.entity_kind;
+        matrix_of_factors.assert_has_entity_kind(entity_kind)?;
```

Comment:
Is this assert needed here?

Reply:
I should perhaps call it `validate` it does panic, it throws. 

And yes it is good to be quite... cautious when it comes to FactorInstance assignment - as we have seen with e.g. the unfortunate Android bug. I made the `assert_has_entity_kind`  validation easy reusable so that we can be cautious. 

Mostly ensuring we don't do anything wrong "internally" - since UI should not allow for this.

Reply:
But it seems redundant here, assert_has_entity will perform the exact same steps as the two line above and then compare the entitiy_kind. Or I am missing something?

Reply:
Thanks, no you are right, thanks! calling `index_agnostic_path_of_all_tx_signing_factor_instances()?` is enough!